provider/ec2/instancetypes.go

```diff
@@ -0,0 +1,348 @@
+		zoneFilter.Values = append(zoneFilter.Values, z.ZoneName)
+	}
+
+	// Query the instance type names for the region and credential in use.
+	var instTypeNames []*string
+	instanceTypeRegions := make(map[string]set.Strings)
+	var token *string
+	for {
+		offeringResults, err := ec2Session.DescribeInstanceTypeOfferings(&ec2.DescribeInstanceTypeOfferingsInput{
+			LocationType: aws.String("availability-zone"),
+			MaxResults:   aws.Int64(maxResults),
+			NextToken:    token,
+			Filters:      []*ec2.Filter{zoneFilter},
+		})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, offering := range offeringResults.InstanceTypeOfferings {
+			if _, ok := instanceTypeRegions[*offering.InstanceType]; !ok {
+				instanceTypeRegions[*offering.InstanceType] = set.NewStrings()
+			}
+			instanceTypeRegions[*offering.InstanceType].Add(*offering.Location)
+			instTypeNames = append(instTypeNames, offering.InstanceType)
+		}
+		token = offeringResults.NextToken
+		if token == nil {
+			break
+		}
+	}
+
+	// Populate the costs for the instance types in use.
+	costs, err := instanceTypeCosts(ec2Session, instTypeNames, zoneNames)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Compose the results.
+	var allInstanceTypes []instances.InstanceType
+	for len(instTypeNames) > 0 {
+		querySize := len(instTypeNames)
+		if querySize > maxResults {
+			querySize = maxResults
+		}
+		page := instTypeNames[0:querySize]
+		instTypeNames = instTypeNames[querySize:]
+
+		instTypeParams := &ec2.DescribeInstanceTypesInput{
+			InstanceTypes: page,
+		}
+		instTypeResults, err := ec2Session.DescribeInstanceTypes(instTypeParams)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, info := range instTypeResults.InstanceTypes {
+			allInstanceTypes = append(
+				allInstanceTypes, convertEC2InstanceType(info, instanceTypeRegions, costs, zoneNames))
+		}
+	}
+
+	if isVPCIDSet(e.ecfg().vpcID()) {
+		return allInstanceTypes, nil
+	}
+	hasDefaultVPC, err := e.hasDefaultVPC(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if hasDefaultVPC {
+		return allInstanceTypes, nil
+	}
+
+	// The region has no default VPC, and the user has not specified
+	// one to use. We filter out any instance types that are not
+	// supported in EC2-Classic.
+	var supportedInstanceTypes []instances.InstanceType
+	for _, instanceType := range allInstanceTypes {
+		if !supportsClassic(instanceType.Name) {
+			continue
+		}
+		supportedInstanceTypes = append(supportedInstanceTypes, instanceType)
+	}
+	return supportedInstanceTypes, nil
+}
+
+func convertEC2InstanceType(
+	info *ec2.InstanceTypeInfo,
+	instanceTypeRegions map[string]set.Strings,
+	costs map[string]uint64,
+	zoneNames []string,
+) instances.InstanceType {
+	instType := instances.InstanceType{
+		Name:       *info.InstanceType,
+		VirtType:   virtType(info),
+		Deprecated: info.CurrentGeneration == nil || !*info.CurrentGeneration,
+		Cost:       costs[*info.InstanceType],
```

Comment:
Is the default cost of 0 treated correctly? We should default unknown costs as max uint64 so they are chosen as a last resort.

Reply:
Other providers default to 0 but then again they don't have costs for anything. 
It's hard to know because in my testing, only 2 instance types didn't get a cost, one being t1.nano, and that's a low cost instance so we'd want to select that if possible.
Maybe using max int is ok